zettarepl/transport/progress_report_mixin.py

```diff
@@ -0,0 +1,68 @@
+# -*- coding=utf-8 -*-
+import logging
+import re
+import threading
+
+from zettarepl.utils.shlex import implode
+
+from .interface import *
+
+logger = logging.getLogger(__name__)
+
+__all__ = ["ProgressReportMixin"]
+
+
+class ProgressReportMixin:
+    def _get_send_shell(self):
+        raise NotImplementedError
+
+    def _zfs_send_can_report_progress(self):
+        send_shell = self._get_send_shell()
+
+        try:
+            send_shell.exec(["zfs", "send", "-V"])
+        except ExecException as e:
+            if "missing snapshot argument" in e.stdout:
+                # Option is supported (patched zfs on FreeNAS)
+                return True
+            else:
+                # invalid option 'V'
+                return False
+        else:
+            return False
+
+    def _wrap_send(self, send):
+        return ["sh", "-c", "(" + implode(send) + " & PID=$!; echo \"zettarepl: zfs send PID is $PID\" 1>&2; "
+                            "wait $PID)"]
+
+    def _start_progress_observer(self):
+        self.stop_progress_observer = threading.Event()
+
+        pid = self.async_exec.head(self._get_zettarepl_pid, 10)
+
+        threading.Thread(daemon=True, name=f"{threading.current_thread().name}.progress_observer",
+                         target=self._progress_observer, args=(pid,)).start()
+
+    def _stop_progress_observer(self):
+        if self.stop_progress_observer:
+            self.stop_progress_observer.set()
+
+    def _get_zettarepl_pid(self, line):
+        m = re.match("zettarepl: zfs send PID is ([0-9]+)", line.strip())
+        if m:
+            return int(m.group(1))
+
+    def _progress_observer(self, pid):
+        send_shell = self._get_send_shell()
+
+        while True:
+            if self.stop_progress_observer.wait(10):
+                return
+
+            s = send_shell.exec(["ps", "-axww", "-o", "command", "-p", str(pid)])
```

Comment:
-p is not for PID in Linux, is it?

Reply:
```
       -p pidlist
              Select by PID.  This selects the processes whose process ID numbers appear in pidlist.  Identical to p and --pid.
```

Reply:
Linux:

william@venon[~]: ps -axww -o command -p 4640|wc -l
777

TrueNAS 12:
root@truenas[~]# ps -axww -o command -p 80918|wc -l
       2

Reply:
@william-gr fixed, without `"-axww"` it works equally correct on FN 11.1, TN 12.0 and Linux